polkadot/runtime/common/src/paras_registrar/mod.rs

```diff
@@ -692,6 +906,138 @@ impl<T: Config> OnNewHead for Pallet<T> {
 	}
 }
 
+impl<T: Config> PreCodeUpgrade for Pallet<T> {
+	/// Ensures that all upgrade-related costs are covered for the specific parachain. Upon success,
+	/// it updates the deposit-related state for the parachain.
+	///
+	/// There are three cases where we do not charge any upgrade costs from the initiator
+	/// of the upgrade:
+	///
+	/// 1. `requirements` is explicitly set to `UpgradeRequirements::SkipRequirements`.
+	///
+	/// 2. System parachains do not pay for upgrade fees.
+	///
+	/// 3. All lease-holding parachains are permitted to do upgrades for free. This is introduced to
+	///    avoid causing a breaking change to the system once para upgrade fees are required.
+	fn pre_code_upgrade(
+		para: ParaId,
+		new_code: ValidationCode,
+		requirements: UpgradeRequirements,
+	) -> Result<Weight, Weight> {
+		let Some(head) = paras::Pallet::<T>::para_head(para) else {
+			return Err(T::DbWeight::get().reads(1))
+		};
+
+		let Some(mut info) = Paras::<T>::get(para) else { return Err(T::DbWeight::get().reads(2)) };
+		let lease_holding = Self::is_parachain(para);
+
+		// The billing account doesn't need to be set for lease holding parachains.
+		if info.billing_account.is_none() && !lease_holding {
```

Comment:
This does not seem to cut it. We don't need a billing account on any free update. I think the code would be easier to reason about, if we removed this check but add a (false, None) match below to handle this.

Reply:
I rearranged the code a bit, should be much easier to follow now.